Refactor rule's test scenarios selection

jan-cerny opened the pull request:
#### Description:

The aim of the PR is to separate code processing rules from code processing test scenarios.

First, we will collect rules to be tested using the `RuleChecker._get_rules_to_test()` method. Then, we will collect a list of test scenarios to be executed using the `RuleChecker._get_scenarios_by_rule_id()` which finds test scenarios for each rule.

#### Rationale:
Rule's test scenarios selection was extended with several features in past year (SSGTS `–scenario` option, templated test scenarios that can be overwritten with rule specific test scenarios) causing the code to be difficult to understand and maintain.

pep8speaks commented:
Hello @jan-cerny! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`tests/unit/ssg_test_suite/test_rule.py`](https://github.com/ComplianceAsCode/content/blob/008a1aab14f53a274b3eb87b86beb7b9c1016b86/tests/unit/ssg_test_suite/test_rule.py):

> [Line 42:1](https://github.com/ComplianceAsCode/content/blob/008a1aab14f53a274b3eb87b86beb7b9c1016b86/tests/unit/ssg_test_suite/test_rule.py#L42): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

jan-cerny commented on tests/ssg_test_suite/rule.py:
```diff
@@ -303,89 +292,86 @@ def _slice_sbr(self, scenarios_by_rule_id, slice_current, slice_total):
+            scenarios.append(scenario)
+        return scenarios
+
+    def _filter_rule_scenarios(self, rule_scenarios):
+        scenarios = []
+        for scenario in rule_scenarios:
+            scenario.override_profile(self.scenarios_profile)
+            if (scenario.matches_regex(self.scenarios_regex) and
+                    scenario.matches_platform(self.benchmark_cpes)):
+                scenarios.append(scenario)
+        return scenarios
```
This says that it filters rule scenarios, but in fact some "filtering" already happened elsewhere, specifically filtering of the templated test scenarios based on the test_config.yml happened in `common.fetch_templated_test_scenarios()` , but that's called from of `_get_all_rule_scenarios`.

So from one point of view it's misleading because:
- `_get_all_rule_scenarios` doesn't get all test scenarios but it gets only those that are allowed by test_config.yml.
- `filter_rule_scenarios` doesn't perform filtering test scenarios base on `test_config.yml` but only based on profile, regex and platform CPE (in other words only based on command line arguments provided by user).

But from the second point of view it's fine because:
- the feature that local test scenarios override templated test scenarios with the same name is now implemented in an easy and elegant way (simply updating the keys within dictionary)
- code handling templated test scenarios is decoupled from code handling not templated test scenarios

jan-cerny replied:
@matejak Do you have any opinion on this subject? It's a question of logical decoupling of responsibilities between different parts of code.

jan-cerny commented:
We have now managed to avoid loading rule.yml files of the rules that won't be tested, only rule.ymls of rules that conform to the user's command line option will be loaded. This speeds up the test by ~15 seconds.

jan-cerny commented on tests/ssg_test_suite/rule.py:
```diff
@@ -261,17 +247,74 @@ def _prepare_environment(self, scenarios_by_rule):
+                    .format(full_rule_id, self.benchmark_id, self.datastream))
+                continue
+
+            # Load the rule itself to check for a template.
+            rule, local_env_yaml = common.load_rule_and_env(
+                dirpath, product_yaml, product)
+
+            # Before we get too far, we wish to search the rule YAML to see if
+            # it is applicable to the current product. If we have a product
+            # and the rule isn't applicable for the product, there's no point
+            # in continuing with the rest of the loading. This should speed up
+            # the loading of the templated tests. Note that we've already
+            # parsed the prodtype into local_env_yaml
+            if product and local_env_yaml['products']:
+                prodtypes = local_env_yaml['products']
+                if "all" not in prodtypes and product not in prodtypes:
+                    continue
+
+            tests_dir = os.path.join(dirpath, "tests")
+            template_name = None
+            if rule.template and rule.template['vars']:
+                template_name = rule.template['name']
```
I noticed that there is a very similar code in the function `template_rule_tests` and there it also loads rules from rule.yml. But the `template_rule_tests`  is called during tarball creation.

I'm not mentioning the code duplication because what is scary is that `template_rule_tests` is called on all rule directories in linux_os tree therefore all rule.yamls (thousands) are parsed and loaded for the second time (!), even if you test a single rule or if you test a rule that doesn't exists. It's done at the moment when we already know which rules and scenarios will be executed.

FTR the call trace just to see where  `template_rule_tests` is called, notice that `self._prepare_environment(scenarios_by_rule_id)` is there which means it's called after we already loaded the rules and scenarios we need.
```
  File "/home/jcerny/work/git/scap-security-guide/tests/test_suite.py", line 462, in <module>
    main()
  File "/home/jcerny/work/git/scap-security-guide/tests/test_suite.py", line 458, in main
    options.func(options)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/rule.py", line 620, in perform_rule_check
    checker.test_target(options.target)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/oscap.py", line 699, in test_target
    self._test_target(target)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/rule.py", line 423, in _test_target
    self._prepare_environment(scenarios_by_rule_id)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/rule.py", line 248, in _prepare_environment
    self.remote_dir = common.send_scripts(self.test_env)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 495, in send_scripts
    archive_file = create_tarball(test_env.product)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 463, in create_tarball
    templated_tests = template_tests(product=product)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 453, in template_tests
    raise exp
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 450, in template_tests
    template_rule_tests(product, product_yaml, template_builder, tmpdir, dirpath)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 386, in template_rule_tests
```

jan-cerny replied:
@matejak  I have transformed this problem into an issue https://github.com/ComplianceAsCode/content/issues/8765 because solving this problem here would make this PR even larger.

matejak replied:
Yes, I completely agree with the description of the issue.

jan-cerny commented:
I have rebased this on master and resolved conflict.

matejak reviewed (changes requested):
I like the PR very much, as it not only makes things clearer, but it also makes them more efficient.

I propose to do additional refactoring focused on naming. I suggest to introduce a concept of id-based vs metadata-based filtering of both rules and scenarios. Then, I suggest to introduce functionality that corresponds to those types of filtering, and to put every other filtering into separate functions.

jan-cerny commented:
I have addressed these problems
- selection of templated test scenarios based on the template used in the tested rules
- split filtering of templated test scenarios based on test_config.yml - there are now multiple different filtering methods
- renaming of the confusing "target" argument

I also have rebased on the latest master.

jan-cerny commented:
I have resolved conflict and rebased on the top of the current master branch.

matejak reviewed (changes requested):
LGTM, I like the PR, and only two small issues are left that are within of the PR scope.

jan-cerny commented:
I have moved the code into _get_rules_to_test and I have removed the "target" argument from test_target methods.

matejak commented:
LGTM, thanks!
